src/Microsoft.AspNetCore.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -307,32 +317,111 @@ private IEnumerable<string> GetAttributeNames(Block tagBlock)
+                            (potentialQuote.Type == HtmlSymbolType.DoubleQuote ||
+                            potentialQuote.Type == HtmlSymbolType.SingleQuote))
+                        {
+                            childOffset = 1;
+                        }
+                    }
+
+                    for (var j = 1; j < childBlock.Children.Count - childOffset; j++)
+                    {
+                        var valueChild = childBlock.Children[j];
+                        if (valueChild.IsBlock)
+                        {
+                            _attributeValueBuilder.Append(InvalidAttributeValueMarker);
+                        }
+                        else
+                        {
+                            var valueChildSpan = (Span)valueChild;
+                            for (var k = 0; k < valueChildSpan.Symbols.Count; k++)
+                            {
+                                _attributeValueBuilder.Append(valueChildSpan.Symbols[k].Content);
+                            }
+                        }
+                    }
                 }
                 else
                 {
                     childSpan = child as Span;
+
+                    var afterEquals = false;
+                    var atValue = false;
+                    var endValueMarker = childSpan.Symbols.Count;
+
+                    // Entire attribute is a string
+                    for (var j = 0; j < endValueMarker; j++)
+                    {
+                        var htmlSymbol = (HtmlSymbol)childSpan.Symbols[j];
+
+                        if (!afterEquals)
+                        {
+                            afterEquals = htmlSymbol.Type == HtmlSymbolType.Equals;
+                            continue;
+                        }
+
+                        if (!atValue)
+                        {
+                            atValue = htmlSymbol.Type != HtmlSymbolType.WhiteSpace &&
+                                htmlSymbol.Type != HtmlSymbolType.NewLine;
+
+                            if (atValue)
+                            {
+                                if (htmlSymbol.Type == HtmlSymbolType.DoubleQuote ||
+                                    htmlSymbol.Type == HtmlSymbolType.SingleQuote)
+                                {
+                                    endValueMarker--;
+                                }
+                                else
+                                {
+                                    // Current symbol is considered the value (unquoted). Add its content to the
+                                    // attribute value builder before we move past it.
+                                    _attributeValueBuilder.Append(htmlSymbol.Content);
+                                }
+                            }
+
+                            continue;
+                        }
+
+                        _attributeValueBuilder.Append(htmlSymbol.Content);
```

Comment:
To confirm: Whitespace after an attribute value (quoted or not) is never part of this `Span`?